### DISCUSSION TITLE
storing data associated with nodes and get nodes by id: best practice in rust?

### DISCUSSION THREAD

**Discussion by chaserhkj** [tree-sitter/tree-sitter#1775] [Q&A] [answered]
Hi I would like to hear some opinion on the best practice to store extra analysis data that are associated with nodes.

Some more context: I am working on a rust project that builds upon `tree-sitter` and doing some sort of static analysis on some custom-language input code. 

So the current way I am doing it is to call `tree-sitter` rust binding and obtain the parsed tree. Then I run the tree thru a traverse algorithm based on the tree cursor and obtained certain analysis data. However, the analysis data is somehow associated with node (e.g. number of function invocation sites is associated with the function definition node) and will need to be accessed multiple times in the next pipelines, but it seems infeasible to directly modify `Node` data structure in rust binding and add the associated fields there.

What is the best practice for this? I was thinking about storing `HashMap<usize, AnalysisResult>`, mapping from node ids to the result of associated analysis. But I feel since node ids are already unique integers, using it as hash map keys means we are hashing it again, which does sound like something wrong to me. Does something like a sparse array makes more sense?

Also, on the other hand, in the associated analysis data, there might be some fields referring to other nodes (e.g. node of function invocation might have associated field pointing to node that defines the function). Since rust does not allow self-referential structs without `Pin`, the only easy way to do this is to store node ids as reference.

However when I looked thru `tree-sitter`'s documentation, it becomes clear that it does not have a `get_node_by_id` method. Does this mean I need to traverse the tree myself to get the node every time I need it? (or to use an index in forms of hash map / sparse array just like the first question)

TLDR: what is the best practice to store data that should be indexed by `tree-sitter` node ids, which are unique integers but not necessarily small?

**Comment by lbguilherme** [comment, marked as answer]
Using `HashMap<usize, AnalysisResult>` seems totally fine for me. Computing the hash of a `usize` is a trivial operation.


I have a similar issue where I want to store complex information on each node. In my case I decided to create my own abstract tree using custom types for each node type and then traverse tree sitter's tree once and building my own tree. Then I can do all operations I want on top of my tree.
I also created operations to update my tree in-place using an updated tree sitter tree after a source code edit.
Only go this route if you really need to.
